micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,480 @@
+        if !DATA_STORE_INITIALIZED.fetch_or(true, Ordering::SeqCst) {
+            let len_per_buffer = unsafe { DATA_STORE.len() } / collector_keys.len();
+            let mut buffers = Vec::new();
+            for i in 0..(collector_keys.len()) {
+                let start_idx = i * len_per_buffer;
+                let end_idx = (i + 1) * len_per_buffer;
+                unsafe {
+                    buffers.push(LocalRb::from_raw_parts(
+                        &mut DATA_STORE[start_idx..end_idx],
+                        0,
+                        0,
+                    ));
+                }
+            }
+            return Ok(Self {
+                buffers,
+                collector_keys,
+            });
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+    ) -> Result<(), DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let encode_len = message.encoded_len();
+        if encode_len + 5 > buffer.vacant_len() {
+            return Err(DataStoreError::DataBufferFull(collector_key, message));
+        }
+        buffer
+            .push(0)
+            .map_err(|_| DataStoreError::DataWriteFailure)?;
+        let len_bytes = (encode_len as u32).to_le_bytes();
+        buffer.push_slice(&len_bytes);
+
+        unsafe {
+            let mut prod = Producer::new(buffer as &LocalRb<u8, &'static mut [MaybeUninit<u8>]>);
+            let (left, right) = prod.free_space_as_slices();
+            let mut chained = BufMut::chain_mut(left, right);
+            message.encode(&mut chained)?;
+            prod.advance(encode_len);
+        }
+
+        Ok(())
+    }
+    fn read_next_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+    ) -> Result<BytesMut, DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+
+        if let Some(&&zero_byte) = buffer.iter().peekable().peek() {
+            if zero_byte != 0 {
+                return Err(DataStoreError::DataIntegrityError);
+            }
+            let _ = buffer.pop();
+        } else {
+            return Ok(BytesMut::with_capacity(0));
+        }
+        let mut encoded_len: [u8; 4] = [0; 4];
+        buffer.pop_slice(&mut encoded_len);
+        let encoded_len = u32::from_le_bytes(encoded_len) as usize;
+        if encoded_len > unsafe { DATA_STORE.len() } {
```

Comment:
I think the bound can be tighter: it can't be longer than any one of the individual ring buffers, right? Though, if you take my suggestion above to not manually encode the length and use the `*_length_delimited` methods this may no longer be possible.